File: 23-merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists == null || lists.length == 0) {
            return null;
        }
        ListNode dummy = new ListNode(-1);
        PriorityQueue<ListNode> heap = new PriorityQueue<ListNode>((node1, node2) -> {
            return node1.val < node2.val ? -1 : 1;
        });
        for (ListNode list : lists) {
            if (list != null) {
                heap.offer(list);      
            }
        }
        ListNode curr = dummy;
        while (!heap.isEmpty()) {
            ListNode node = heap.poll();
            curr.next = node;
            curr = curr.next;
            if (node.next != null) {
                heap.offer(node.next);
            }
        }
        curr.next = null;
        return dummy.next;
    }
}

File: python3/23.merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        min_heap = []
        for index, node in enumerate(lists):
            if node is not None:
                heapq.heappush(min_heap, (node.val, index, node));
        
        dummy = ListNode()
        curr = dummy
        while min_heap:
            _, index, min_node = heapq.heappop(min_heap)
            curr.next = min_node
            curr = curr.next
            
            if min_node.next is not None:
                next_node = min_node.next
                min_node.next = None
                
                heapq.heappush(min_heap, (next_node.val, index, next_node))
                
        return dummy.next
